### DISCUSSION TITLE
In vim9script, char2nr() does not return correct values for UTF8 characters

### DISCUSSION THREAD

**Discussion by kennypete** [vim/vim#15325] [Q&A] [answered]
### Steps to reproduce

The following vimscript will produce correct output for the UTF8 character Ā:

```vim
function g:BytesUTF8()
  echo char2nr("Ā"[0], 0) .. ' ' .. char2nr("Ā"[1], 0)
  echo char2nr("Ā"[0], 1) .. ' ' .. char2nr("Ā"[1], 1)
endfunction
```

Output echoed:
```
196 128
196 128
```

The following vim9script should do the same thing:

```vim
vim9script
def g:BytesUTF8vim9()
  echo char2nr("Ā"[0], 0) .. ' ' .. char2nr("Ā"[1], 0)
  echo char2nr("Ā"[0], 1) .. ' ' .. char2nr("Ā"[1], 1)
enddef
```

It doesn't. Its echoed output is:

```
256 0
256 0
```

In the vim9script function, what's being returned is the Unicode code point and a zero, not what should be returned, i.e., the UTF8 bytes, like the _identical_ lines do in vimscript.

### Expected behaviour

The output of `char2nr()` should be the same, regardless of whether the script is vimscript or vim9script. Also note, `:echo char2nr("Ā"[0]) .. ' ' .. char2nr("Ā"[1])` in cmdline mode similarly produces correct `196 128` output.

### Version of Vim

9.1.90

### Environment

The same result happens in Windows 11 (GUI or console) or WSL Debian. It also happens back to (at least) 8.2.1584, whether run using `vim --clean`, etc.

**Comment by errael** [comment]
I don't know if this is relevant to the issue you're experiencing, but indexing a string is different between `legacy` and `vim9script`.

`:he vim9-string-index` says
```
Indexing a string with [idx] or taking a slice with [idx : idx] uses character
indexes instead of byte indexes.  Composing characters are included.
```

**Comment by zeertzjq** [comment, marked as answer]
And instead you can use `strpart()` for byte indexing:
```vim
vim9script
def g:BytesUTF8vim9()
  echo char2nr("Ā"->strpart(0, 1), 0) .. ' ' .. char2nr("Ā"->strpart(1, 1), 0)
  echo char2nr("Ā"->strpart(0, 1), 1) .. ' ' .. char2nr("Ā"->strpart(1, 1), 1)
enddef
```

**Reply by kennypete** [reply]
I'll check out what you both have replied with, though may look to submit a CR for builtin.txt because there does not appear to be
any suggestion in :h char2nr() that the function is not forwards compatible in vim9script? (The examples provided you'd expect should work in both unless it's noted otherwise?)

**Reply by kennypete** [reply]
Thanks, that works well. I wonder why the change to character indexes rather than byte indexes was made to `char2nr()`? It looks like it could have retained the same functionality. I can't write C, but reading `f_char2nr`, in evalfunc.c, it looks like it may have been possible to have it behave as it does in vimscript / the command line. No matter, I presume there was a good reason not to do that and, although the vim9script approach is more complicated, at least the capability is still there.

That change, combined with "Local variables will not be visible to string evaluation" (under Limitations in [vim9.txt](https://vimhelp.org/vim9.txt.html#fast-functions) - also not so easy to find, IMHO), really had me stumped. That is, not being able to just change the `c` to `c:string` and remove the `a:` instances from something like:

```vim
function! g:C(c)
  echo join(map(range(strlen(a:c)), 'a:c'), ' ')
endfunction

```
... and instead needing to do this:

```vim
vim9script
def! g:C9(c: string)
  echo join(range(strlen(c))->map((_, v) => c), ' ')
enddef
```

Back to `char2nr()`. Reading [char2nr()](https://vimhelp.org/builtin.txt.html#char2nr%28%29) help, I think it needs a caveat on `char2nr("á"[0])  returns 195`. Unrelated, it also seems to have an error in `Returns 0 if {string} is not a String` because `char2nr(1)` returns `49`, for example. And other non-string content results in an error (e.g., `char2nr([1, 2])` is an E730 error, not 0).
 
Is there any vimhelp writing guide? My first crack at what could be written after the line, `char2nr("á"[0])  returns 195`:

```
<			Note: In |vim9script| 195 is not returned because
			it uses character indexes instead of byte indexes.
			Consequently, "á"[0] returns 225.
			To return Number values of bytes, use |strpart()|: >
			vim9script
			def g:C2example9()
			  echo char2nr("á"[0])			| # returns 225
			  echo char2nr("á"[1])			| # returns 0
			  echo char2nr("á"->strpart(0, 1))	| # returns 195
			  echo char2nr("á"->strpart(1, 1))	| # returns 161
			enddef
<			Refer to |vim9-string-index| for more information.
```

**Reply by zeertzjq** [reply]
`char2nr()` does not use indexes. It uses the first character of the string provided to it. The different you are observing here is the behavior described in `:h vim9-string-index`. You can see that by echoing the string passed to `char2nr()`:
```vim
vim9script
echo "Ā"[0]  # prints "Ā"
echo "Ā"[1]  # prints ""
```

**Reply by kennypete** [reply]
> `char2nr()` does not use indexes. 

That's inconsistent with `:h vim9-string-index` and what is returned by `char2nr()`. The help says, "...uses character indexes instead of byte indexes", and it happens in practice:

```vim
echo char2nr('Ā☺️'[1]) | " prints 128 (decimal value of byte index 1: 80 from the UTF-8, C2 80, of Ā)
```
whereas:
```vim
vim9script
echo char2nr('Ā☺️'[1])   # prints 9786 (decimal value of character index 1, U+263A, being ☺️)
```

**Reply by zeertzjq** [reply]
The index isn't being passed to `char2nr()`. It is only used in the string indexing.

**Reply by kennypete** [reply]
> The index isn't being passed to `char2nr()`. It is only used in the string indexing.

Sorry, thanks for your patience. That's sunk in now. Yes, `char2nr()` is doing the same thing. It's the input to it -- byte versus character indexing -- that's the difference.

Regardless, the function is not really _char2nr_, it's more like _firsthex2dec_ where, in vimscript, _firsthex_ is the first UTF-8 byte and, in vim9script, _firsthex_ is the first Unicode character.
